core/src/Kokkos_ReductionIdentity.hpp

```diff
@@ -22,407 +22,71 @@
-    return UINT_MAX;
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned int bor() {
-    return static_cast<unsigned int>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned int band() {
-    return ~static_cast<unsigned int>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned int lor() {
-    return static_cast<unsigned int>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned int land() {
-    return static_cast<unsigned int>(1);
-  }
-};
-
-template <>
-struct reduction_identity<unsigned long> {
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long sum() {
-    return static_cast<unsigned long>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long prod() {
-    return static_cast<unsigned long>(1);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long max() {
-    return static_cast<unsigned long>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long min() {
-    return ULONG_MAX;
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long bor() {
-    return static_cast<unsigned long>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long band() {
-    return ~static_cast<unsigned long>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long lor() {
-    return static_cast<unsigned long>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long land() {
-    return static_cast<unsigned long>(1);
-  }
-};
-
-template <>
-struct reduction_identity<unsigned long long> {
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long sum() {
-    return static_cast<unsigned long long>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long prod() {
-    return static_cast<unsigned long long>(1);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long max() {
-    return static_cast<unsigned long long>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long min() {
-    return ULLONG_MAX;
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long bor() {
-    return static_cast<unsigned long long>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long band() {
-    return ~static_cast<unsigned long long>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long lor() {
-    return static_cast<unsigned long long>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned long long land() {
-    return static_cast<unsigned long long>(1);
-  }
+struct reduction_identity;
+
+template <typename Integral>
+  requires(std::integral<Integral>)
+struct reduction_identity<Integral> {
+  KOKKOS_FUNCTION constexpr static Integral sum() noexcept { return 0; }
+  KOKKOS_FUNCTION constexpr static Integral prod() noexcept { return 1; }
+  KOKKOS_FUNCTION constexpr static Integral max() noexcept {
+    return std::numeric_limits<Integral>::min();
```

Comment:
I'd expect calling a host function here to be problematic with `Cuda` and `HIP`. In my opinion, we should rather use our numeric limits header here (as proposed in https://github.com/masterleinad/kokkos/commit/6f8920a8826990604be32ec3d14e54db76dab88e).

Reply:
Relying on numeric limits would allow us to also deal with the half types here (with some minor changes).

Reply:
It is about trying to bring sanity in how we architecture the codebase.